### PR TITLE
Propose to fix #650

### LINKED ISSUE #650: ReduceLROnPlateau
Dear all,

first of all, I love this library. 

The `ReduceLROnPlateau` is not working when I call `trail.evaluate`.

```python
...
    trial = Trial(cnn, optimizer, loss, metrics=['acc', 'loss'],
                  callbacks=[
                    #   CometCallback(experiment),
                      ReduceLROnPlateau(monitor='val_loss',
                                        factor=0.1, patience=5),
                    #   EarlyStopping(monitor='val_acc', patience=5, mode='max'),
                    #   CSVLogger('history.csv'),
                    #   ModelCheckpoint('best.pt', monitor='val_acc', mode='max')
    ]).to(device)
    trial.with_generators(train_generator=train_dl,
                          val_generator=val_dl, test_generator=test_dl)
    # history = trial.run(params['epochs'], verbose=1)
    preds = trial.evaluate(data_key=torchbearer.TEST_DATA)
```

error: 

```
0/1(e): 100%|███████████████████████████████████| 1/1 [00:00<00:00,  2.18it/s, test_acc=0.4667, test_loss=0.6516]
Traceback (most recent call last):
  File "c:/Users/Francesco/Documents/PyTorch-Deep-Learning-Template/main.py", line 62, in <module>
    preds = trial.evaluate(data_key=torchbearer.TEST_DATA)
  File "C:\Users\Francesco\Anaconda3\envs\dl\lib\site-packages\torchbearer\trial.py", line 298, in wrapper
    res = func(self, *args, **kwargs)
  File "C:\Users\Francesco\Anaconda3\envs\dl\lib\site-packages\torchbearer\trial.py", line 133, in wrapper
    res = func(self, *args, **kwargs)
  File "C:\Users\Francesco\Anaconda3\envs\dl\lib\site-packages\torchbearer\trial.py", line 1131, in evaluate
    state[torchbearer.CALLBACK_LIST].on_end_epoch(state)
  File "C:\Users\Francesco\Anaconda3\envs\dl\lib\site-packages\torchbearer\callbacks\callbacks.py", line 221, in on_end_epoch
    self._for_list(lambda callback: callback.on_end_epoch(state))
  File "C:\Users\Francesco\Anaconda3\envs\dl\lib\site-packages\torchbearer\trial.py", line 105, in _for_list
    function(self.callback_list)
  File "C:\Users\Francesco\Anaconda3\envs\dl\lib\site-packages\torchbearer\callbacks\callbacks.py", line 221, in <lambda>
    self._for_list(lambda callback: callback.on_end_epoch(state))
  File "C:\Users\Francesco\Anaconda3\envs\dl\lib\site-packages\torchbearer\callbacks\callbacks.py", line 221, in on_end_epoch
    self._for_list(lambda callback: callback.on_end_epoch(state))
  File "C:\Users\Francesco\Anaconda3\envs\dl\lib\site-packages\torchbearer\callbacks\callbacks.py", line 66, in _for_list
    function(callback)
  File "C:\Users\Francesco\Anaconda3\envs\dl\lib\site-packages\torchbearer\callbacks\callbacks.py", line 221, in <lambda>
    self._for_list(lambda callback: callback.on_end_epoch(state))
  File "C:\Users\Francesco\Anaconda3\envs\dl\lib\site-packages\torchbearer\callbacks\torch_scheduler.py", line 32, in on_end_epoch
    self._scheduler.step(state[torchbearer.METRICS][self._monitor], epoch=state[torchbearer.EPOCH])
KeyError: 'val_loss'
```

Probably you need to disable the callback when evaluating (or just checking if the monitored metrics `is in state['metrics']`.

Thank you!

Best Regards,

Francesco Saverio

### PR DESCRIPTION
**Pull request by FrancescoSaverioZuppichini**
We can check if `self._monitor` is present inside the given metrics dict

**No tested**

### CODE REVIEW

**Comment by MattPainter01** [comment]
Thanks for the PR, looks good! 

I've had a quick look at the build errors and it looks like the current version of torchvision doesn't support pillow version 7. Your build errors should be fixed when they release a new torchvision version this week. 

One quick thing, If you could add some warnings saying something along the lines of "Monitor key [KEY] was not found by the scheduler" then I'd be happy to merge it, although maybe after the torchvision release so the tests all pass. These warnings will probably require an extra test or two to be covered, testing warnings is a bit awkward but we've done it in other places I believe. I can add them afterwards if it's easier.

**Comment by FrancescoSaverioZuppichini** [comment]
@MattPainter01 Sure, I will add them after work. We should also add a test or two to the Callback to be sure everything works.

**Comment by FrancescoSaverioZuppichini** [comment]
@MattPainter01  added warnings.

File: torchbearer/callbacks/torch_scheduler.py [hunk 1]
```diff
@@ -21,15 +21,22 @@ def on_sample(self, state):
...
         if self._step_on_batch and self._monitor is not None:
-            self._scheduler.step(state[torchbearer.METRICS][self._monitor])
+            if self._monitor in state[torchbearer.METRICS]:
+                self._scheduler.step(state[torchbearer.METRICS][self._monitor])
 
     def on_start_training(self, state):
         if not self._step_on_batch and self._monitor is None:
             self._scheduler.step(epoch=state[torchbearer.EPOCH])
+        else:
+            warnings.warn(f"Monitor key [{self._monitor}] was not found by the scheduler.", Warning)
```
Comment (MattPainter01) [inline torchbearer/callbacks/torch_scheduler.py]: - I think this warning should go after line 25, at the moment it's just emitting a warning whenever we're not stepping on batch. 
- We should probably also use string formatting that works on both python 2 and 3 for the moment, even if python 2 is unsupported these days. 
I believe ``"Monitor key [{}] was not found by the scheduler.".format(self._monitor)`` works on both versions.

**Comment by FrancescoSaverioZuppichini** [comment]
@MattPainter01 I have updated the code!

**Comment by MattPainter01** [comment]
Great, thanks! 

It seems like torchvision still haven't put out a new release so I'll write some tests based on an older version of torchvision/pillow and add them here.

**Comment by FrancescoSaverioZuppichini** [comment]
@MattPainter01 good to hear! I cannot wait to start using torchbearer in my DL workflow :)

**Comment by MattPainter01** [comment]
Marged, thanks for the help @FrancescoSaverioZuppichini!

**Pull request merged**